File: web/src/api/deployment.ts
import client from './client';

export interface EnvVar {
  id: string;
  project_id: string;
  key: string;
  value: string;
  is_secret: boolean;
  created_at: string;
  updated_at: string;
}

export interface VPSTarget {
  id: string;
  project_id: string;
  name: string;
  host: string;
  port: number;
  username: string;
  auth_type: 'key' | 'password';
  deploy_path: string;
  pre_deploy_cmd: string;
  deploy_cmd: string;
  post_deploy_cmd: string;
  status: 'idle' | 'deploying' | 'success' | 'failed';
  last_deployed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface DeploymentRun {
  id: string;
  vps_target_id: string;
  project_id: string;
  status: 'running' | 'success' | 'failed';
  triggered_by?: string;
  log_output: string;
  started_at: string;
  finished_at?: string;
  created_at: string;
  target?: VPSTarget;
}

export interface TargetInput {
  name: string;
  host: string;
  port?: number;
  username: string;
  auth_type: 'key' | 'password';
  ssh_key?: string;
  ssh_password?: string;
  deploy_path: string;
  pre_deploy_cmd?: string;
  deploy_cmd: string;
  post_deploy_cmd?: string;
}

export const deploymentApi = {
  // Env vars
  listEnvVars: (projectId: string) =>
    client.get<{ env_vars: EnvVar[] }>(`/projects/${projectId}/env`),

  setEnvVar: (projectId: string, key: string, value: string, isSecret = false) =>
    client.post<{ env_var: EnvVar }>(`/projects/${projectId}/env`, {
      key,
      value,
      is_secret: isSecret,
    }),

  deleteEnvVar: (projectId: string, envId: string) =>
    client.delete(`/projects/${projectId}/env/${envId}`),

  // VPS targets
  listTargets: (projectId: string) =>
    client.get<{ targets: VPSTarget[] }>(`/projects/${projectId}/deploy`),

  createTarget: (projectId: string, input: TargetInput) =>
    client.post<{ target: VPSTarget }>(`/projects/${projectId}/deploy`, input),

  updateTarget: (projectId: string, targetId: string, input: TargetInput) =>
    client.put<{ target: VPSTarget }>(`/projects/${projectId}/deploy/${targetId}`, input),

  deleteTarget: (projectId: string, targetId: string) =>
    client.delete(`/projects/${projectId}/deploy/${targetId}`),

  // Deploy
  deploy: (projectId: string, targetId: string) =>
    client.post<{ run: DeploymentRun }>(`/projects/${projectId}/deploy/${targetId}/trigger`),

  // Runs
  listRuns: (projectId: string, targetId?: string, limit = 20) => {
    const base = targetId
      ? `/projects/${projectId}/deploy/${targetId}/runs`
      : `/projects/${projectId}/deploy/runs`;
    return client.get<{ runs: DeploymentRun[] }>(`${base}?limit=${limit}`);
  },

  getRun: (projectId: string, runId: string) =>
    client.get<{ run: DeploymentRun }>(`/projects/${projectId}/runs/${runId}`),
};

File: internal/models/deployment.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ProjectEnvVar stores encrypted env vars/secrets for a project.
type ProjectEnvVar struct {
	ID        uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	ProjectID uuid.UUID `gorm:"type:uuid;not null;index" json:"project_id"`
	Key       string    `gorm:"size:255;not null" json:"key"`
	Value     string    `gorm:"type:text;not null" json:"-"` // AES-encrypted, never serialized to clients
	IsSecret  bool      `gorm:"default:false" json:"is_secret"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (e *ProjectEnvVar) BeforeCreate(tx *gorm.DB) error {
	if e.ID == uuid.Nil {
		e.ID = uuid.New()
	}
	return nil
}

// VPSTarget is a configured SSH deployment target for a project.
type VPSTarget struct {
	ID             uuid.UUID  `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	ProjectID      uuid.UUID  `gorm:"type:uuid;not null;index" json:"project_id"`
	Name           string     `gorm:"size:255;not null" json:"name"`
	Host           string     `gorm:"size:255;not null" json:"host"`
	Port           int        `gorm:"default:22" json:"port"`
	Username       string     `gorm:"size:255;not null" json:"username"`
	AuthType       string     `gorm:"size:20;default:key" json:"auth_type"` // "key" or "password"
	SSHKey         string     `gorm:"type:text" json:"-"`                   // AES-encrypted PEM private key
	SSHPassword    string     `gorm:"type:text" json:"-"`                   // AES-encrypted password
	DeployPath     string     `gorm:"size:500;not null" json:"deploy_path"`
	PreDeployCmd   string     `gorm:"type:text" json:"pre_deploy_cmd"`
	DeployCmd      string     `gorm:"type:text;not null" json:"deploy_cmd"`
	PostDeployCmd  string     `gorm:"type:text" json:"post_deploy_cmd"`
	Status         string     `gorm:"size:20;default:idle" json:"status"` // idle | deploying | success | failed
	LastDeployedAt *time.Time `json:"last_deployed_at,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
}

func (v *VPSTarget) BeforeCreate(tx *gorm.DB) error {
	if v.ID == uuid.Nil {
		v.ID = uuid.New()
	}
	return nil
}

// DeploymentRun records a single deployment execution.
type DeploymentRun struct {
	ID          uuid.UUID  `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	VPSTargetID uuid.UUID  `gorm:"type:uuid;not null;index" json:"vps_target_id"`
	ProjectID   uuid.UUID  `gorm:"type:uuid;not null;index" json:"project_id"`
	Status      string     `gorm:"size:20;default:running" json:"status"` // running | success | failed
	TriggeredBy *uuid.UUID `gorm:"type:uuid" json:"triggered_by,omitempty"`
	LogOutput   string     `gorm:"type:text;default:''" json:"log_output"`
	StartedAt   time.Time  `gorm:"default:NOW()" json:"started_at"`
	FinishedAt  *time.Time `json:"finished_at,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`

	VPSTarget *VPSTarget `gorm:"foreignKey:VPSTargetID" json:"target,omitempty"`
}

func (d *DeploymentRun) BeforeCreate(tx *gorm.DB) error {
	if d.ID == uuid.Nil {
		d.ID = uuid.New()
	}
	return nil
}
